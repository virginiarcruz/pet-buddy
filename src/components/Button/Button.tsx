import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'
import { ButtonContainer, Icon } from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  as?: React.ElementType
  children?: React.ReactNode
} & ButtonTypes

const Button = ({ children, ...props }: ButtonProps) => (
  <ButtonContainer {...props}>
    <span>{children}</span>
    <Icon />
  </ButtonContainer>
)

export default Button
