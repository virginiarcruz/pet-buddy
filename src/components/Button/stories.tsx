import { Story, Meta } from '@storybook/react/types-6-0'
import Button from '.'

export default {
  title: 'Button',
  component: Button
} as Meta

export const Default: Story = (args) => <Button {...args}> Details </Button>
export const asLink: Story = (args) => <Button {...args}> Details </Button>
asLink.args = {
  as: 'a',
  href: '/link'
}
