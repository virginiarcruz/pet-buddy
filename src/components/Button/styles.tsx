import styled, { css, DefaultTheme } from 'styled-components'
import IconButton from './assets/arrow.svg'

import { ButtonProps } from './Button'

const buttonModifiers = {
  link: (theme: DefaultTheme) => css`
    background: none;
    padding: 0;
    color: ${theme.color.blue};
    text-decoration: none;
    svg {
      path {
        fill: ${theme.color.blue};
      }
    }
  `
}

export const ButtonContainer = styled.button<ButtonProps>`
  ${({ theme, as }) => css`
    border: 0;
    border-radius: ${theme.border.radius.small};
    background-color: ${theme.color.blue};
    color: ${theme.color.white};
    font-family: ${theme.font.family};
    font-size: ${theme.font.size.medium};
    font-weight: ${theme.font.bold};
    line-height: 19px;
    padding: 16px 99px;
    max-width: 263px;
    cursor: pointer;

    display: flex;
    align-items: center;

    ${!!as && buttonModifiers.link(theme)};
  `}
`

export const Icon = styled(IconButton)`
  ${({ theme }) => css`
    margin-left: 4px;
    width: 7px;
    height: 12px;
    path {
      fill: ${theme.color.white};
    }
  `}
`
