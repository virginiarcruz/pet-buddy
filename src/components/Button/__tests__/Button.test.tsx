import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import Button from '../Button'

describe('<Button />', () => {
  it('should render the <Button />', () => {
    const { container } = renderWithTheme(<Button> my button</Button>)

    expect(container).toMatchSnapshot()

    expect(screen.getByRole('button', { name: /my button/i })).toHaveStyle({
      'max-width': '263px',
      padding: '16px 99px',
      'font-size': '16px'
    })
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        my button
      </Button>
    )

    expect(screen.getByRole('link', { name: /my button/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
