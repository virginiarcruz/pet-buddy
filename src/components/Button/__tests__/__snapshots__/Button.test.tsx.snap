// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the <Button /> 1`] = `
.c0 {
  border: 0;
  border-radius: 8px;
  background-color: #002FD6;
  color: #FFFFFF;
  font-family: 'Inter',-apple-system,sans-serif;
  font-size: 16px;
  font-weight: 600;
  line-height: 19px;
  padding: 16px 99px;
  max-width: 263px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  margin-left: 4px;
  width: 7px;
  height: 12px;
}

.c1 path {
  fill: #FFFFFF;
}

<div>
  <button
    class="c0"
  >
    <span>
       my button
    </span>
    <svg
      class="c1"
      viewBox="0 0 7 12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.821 6.454a.665.665 0 0 0 0-.908L1.687.186c-.264-.248-.661-.248-.9 0L.18.818c-.239.247-.239.66 0 .935l4.075 4.233L.18 10.247c-.239.275-.239.688 0 .935l.608.632c.239.248.636.248.9 0l5.134-5.36z"
      />
    </svg>
  </button>
</div>
`;
